mcf-browser/src/back-end/graph.js

```diff
@@ -0,0 +1,318 @@
+  /**
+   * Whether triples from the remote Data Commons Knowledge Graph have already
+   * been fetched.
+   * @type {boolean}
+   */
+  alreadyFetched;
+  /**
+   * Whether the node exists in the Data Commons Knowledge Graph.
+   * @type {boolean}
+   */
+  existsInKG;
+  /**
+   * Dcid of the node. Set only if a remote id is referred to in a local file or
+   * pulled from the Data Commons Knowledge Graph.
+   * @type {?string}
+   */
+  dcid;
+  /**
+   * The head of a linked list of Assertion objects representing the outgoing
+   * triples of the Node object.
+   * @type {?Assertion}
+   */
+  assertions;
+  /**
+   * The head of a linked list of Assertion objects representing the incoming
+   * triples of the Node object.
+   * @type {?Assertion}
+   */
+  invAssertions;
+
+  /**
+   * Create a Node based on a given id.
+   * @param {string} id The id of the node to create.
+   */
+  constructor(id) {
+    this.localId = id;
+    this.alreadyFetched = false;
+    this.existsInKG = false;
+    this.dcid = null;
+    this.assertions = null;
+    this.invAssertions = null;
+  }
+
+  /**
+   * Indicates if a given object is an instance of Node class.
+   * @param {Object} obj The object to check.
+   * @return {boolean} True if the object is an instance of Node.
+   */
+  static isNode(obj) { return obj instanceof Node; }
+
+  /**
+   * Returns a node with the given ID, creates a new node if shouldCreate is
+   * true, or null if the node does not exist and a node should not be created.
+   *
+   * @param {string} id The id of the node to find.
+   * @param {boolean} shouldCreate True if a new Node should be created if it
+   *     does not already exist.
+   * @return {Node|null} The found node if it exists or is created.
+   */
+  static getNode(id, shouldCreate) {
+    const existing = Node.nodeHash[id];
+    if (existing) {
+      return existing;
+    }
+
+    if (shouldCreate) {
+      const newNode = new Node(id);
+      Node.nodeHash[id] = newNode;
+      return newNode;
+    }
+    return null;
+  }
+
+  /**
+   * Returns the reference to the node that is displayed in browser. If the
+   * node has a dcid, then the dcid will be displayed. If the node's local id
+   * is different, then the local id is also displayed.
+   * Ex: <dcid> [l:<localId>]
+   * @return {string} The reference to the node to be displayed.
+   */
+  getRef() {
+    const dcidRef = this.dcid ? this.dcid : '';
+    let localRef = '';
+
+    if (!this.dcid || this.dcid !== this.localId) {
+      localRef = '[l:' + this.localId + ']';
+    }
+    return [ dcidRef, localRef ].join(' ').trim();
+  }
+
+  /**
+   * Sets the dcid of Node object. Checks if a separate node based on the dcid
+   * already exists. If remote node exists, then the remote node is absorbed by
+   * current node via mergeNode() method.
+   *
+   * @param {string} dcid The dcid to be added to the Node object.
+   */
+  setDCID(dcid) {
+    const remote = Node.nodeHash[dcid];
+    if (remote) {
```

Comment:
Is it more accurate to check for remote != this ? e.g., calling setDCID twice on the same node

Reply:
should I throw some sort of error or warning if the user is trying to set a dcid on a node that already has a different dcid? for example. if there is an error in mcf similar to
Node: foo
dcid: remoteFoo
dcid: alsoRemoteFoo

the merge function checks to make sure a node is not being merged with itself, should this check be performed in setDCID before calling the merge as well?